C10D: Added TCPStore to support C10D store interface

teng-li opened the pull request:
Added TCP Store as another Store for the C10D interface. 

For more details on the C10D Store interface, refer: https://github.com/pytorch/pytorch/issues/7434

A few things done in the PR:

(1) TCP Store can be initialized either as a master who will launch a daemon thread or a slave who will connect to the master's store daemon thread.
(2) TCP Store supports all the interface functions for Store.h
(3) Refactored some of the TCP helper functions into Utils.hoo/cpp
(4) Refactored test to some common folders.

Test Plan:
```
~/pytorch/torch/lib/c10d$ bin/test.sh
+ mkdir -p build
+ cd build
+ cmake ../
-- Configuring done
-- Generating done
-- Build files have been written to: /private/home/tengli/pytorch/torch/lib/c10d/build
+ make all test
[ 11%] Building CXX object CMakeFiles/store.dir/Utils.cpp.o
[ 22%] Building CXX object CMakeFiles/store.dir/Store.cpp.o
[ 33%] Building CXX object CMakeFiles/store.dir/FileStore.cpp.o
[ 44%] Building CXX object CMakeFiles/store.dir/TcpStore.cpp.o
[ 55%] Linking CXX static library libstore.a
[ 55%] Built target store
Scanning dependencies of target TcpStoreTest
[ 66%] Building CXX object test/CMakeFiles/TcpStoreTest.dir/TcpStoreTest.cpp.o
[ 77%] Linking CXX executable TcpStoreTest
[ 77%] Built target TcpStoreTest
[ 88%] Building CXX object test/CMakeFiles/FileStoreTest.dir/FileStoreTest.cpp.o
[100%] Linking CXX executable FileStoreTest
[100%] Built target FileStoreTest
Running tests...
Test project /private/home/tengli/pytorch/torch/lib/c10d/build
    Start 1: FileStoreTest
1/2 Test #1: FileStoreTest ....................   Passed    0.03 sec
    Start 2: TcpStoreTest
2/2 Test #2: TcpStoreTest .....................   Passed    0.69 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =   0.74 sec
```

teng-li committed:
C10D: Added TCPStore to support C10D store interface

apaszke commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -0,0 +1,282 @@
+int connect(const std::string& address,
+            PortType port,
+            bool wait,
+            int timeout) {
```
We might want to start using `std::chrono` structs for this

apaszke commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -0,0 +1,192 @@
+
+    bytesToReceive -= bytesReceived;
+    currentBytes += bytesReceived;
+  }
```
We did get some `read`/`write` wrappers in @pietern's PR. We might want to deduplicate those.

apaszke replied:
I can't see the difference. Both are using `read`/`write` to receive or put data into file descriptors.

teng-li replied:
they are different, one is using ::read, one is using ::recv. And we need to take care of the case when byteReceived == 0, we throw an error, that's not the logic in the File class's case

pietern replied:
What is the reason for send/recv vs read/write? Do we need the additional controls that they give us? No need to change this, but want to understand why not read/write.

teng-li replied:
They should be the same. Used send/recv since the original THD util uses them.  The only difference is that here we check 0 bytes and throw an exception.

apaszke commented on torch/lib/c10d/test/TcpStoreTest.cpp:
```diff
@@ -0,0 +1,63 @@
+              c10d::test::set(*slaveStores[i], key, val);
+              c10d::test::check(*slaveStores[i], key, val);
+            }
+          })));
```
Can we also have a check for reading values written by other threads? Just have a barrier on the `Semaphore`

teng-li replied:
Added

apaszke reviewed (commented):
It looks like most of the code is coming straight from THD, is that right? Can you please post diffs comparing them to the original files? That would make reviewing much easier.

teng-li commented:
@apaszke Util.cpp/hpp are mostly from THD with some refactoring.  TcpStore.hpp/cpp are based on architecture of Gloo store, but the API implementation needs some changes to support timed wait, add, check functions that are defined in the base store class. I sent you a diff file internally.

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,274 @@
+
+{
+  if (isServer_) {
+    // Openning up the listening socket
```
Typo

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,274 @@
+    return false;
+  } else {
+    throw std::runtime_error("stop_waiting or keep_waiting response expected");
+  }
```
The server will return another QueryType::STOP_WAITING when it is done waiting. This will be unexpected by whatever function is called next. I think it would be easier to split "check" and "wait" into two paths, one that is blocking and one that is non blocking.

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,274 @@
+ * or, in the case of wait
+ * type of query | number of args | size of arg1 | arg1 | ...
+ */
+void TcpStoreDaemon::query(RankType rank) {
```
Same comment about rank. I think we can just pass the fd here.

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,274 @@
+void TcpStore::set(const std::string& key, const std::vector<uint8_t>& data) {
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::SET);
+  tcputil::sendString(storeSocket_, key, true);
+  tcputil::sendVector<uint8_t>(storeSocket_, data);
```
What do you think of making every command request/response to make the interaction pattern predictable?

pietern commented on torch/lib/c10d/test/CMakeLists.txt:
```diff
@@ -13,4 +14,5 @@ foreach(test_src ${test_srcs})
   target_include_directories(${test_name} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/..)
   target_link_libraries(${test_name} ${test_libraries})
   add_test(NAME ${test_name} COMMAND $<TARGET_FILE:${test_name}>)
+
```
Remove

pietern reviewed (commented):
Nice work @teng-li!

General comments:
* Can we name it TCPStore instead?
* There is inconsistency in style, w.r.t. double newlines, braces on newline vs end of line, function argument indentation, namespace comments.
* A few comments below talk about this: I think we should not let the exception propagation determine shutdown of the daemon thread. It will happen one node dies for whatever reason and this will immediately propagate everywhere. Instead it should just go away, be removed from the fd vector, and the daemon should keep running.

teng-li committed:
Used pipe to terminate the store daemon and addressed all comments

Please see: https://github.com/pytorch/pytorch/issues/7434

teng-li commented:
@pietern Changed the file name to TCPStore, but somehow the new commit cannot see the diff :(,  I can revert the name back so that you can see the diff and later rename it.

Anyway, addressed all of your comments, like properly shutdown the daemon using pipes etc.

@apaszke  addressed your comments too,

apaszke commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,326 @@
+      ret = false;
+      it++;
+    } else {
+      it = keys.erase(it);
```
It looks like this code codes a lot of unnecessary things. I think the whole problem could be solved as (without modifying the vector):
```cpp 
return std::all_of(keys.begin(), keys.end(),
                   [](const std::string& s) { return tcpStore_.count(s) > 0; });
```

teng-li replied:
Done

apaszke commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,326 @@
+    }
+    /* sleep override */
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  }
```
Perhaps this should just become the default implementation of `wait` for different stores? I think we're using the same strategy both in here and in the file one. Wdyt @pietern?

pietern replied:
We can do this as a stop gap but I would rather make this wake up to activity. For the TCPStore this means issuing a wait requests that returns as soon as it is fulfilled (the waiting part is done on the server side).

apaszke replied:
That’s what the impl used to in THD (and before the last commit), so we might want to revert this part.

pietern replied:
The commit before last risked running out of sync between server and client. The notify/wakeup part can be reverted, but should hold on to the socket that expects a message for wakeup, not the keep_waiting/stop_waiting part.

apaszke replied:
Ok, my bad. That must have been changed from THD, [which only has the `STOP_WAITING` command](https://github.com/pytorch/pytorch/blob/master/torch/lib/THD/base/data_channels/Store.cpp#L16).

teng-li replied:
Ok, I changed it back.

apaszke commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -0,0 +1,196 @@
+}
+
+// send a string's length and data
+inline void sendString(int socket,
```
At some point we should just move to an overloaded `send` and `recv` function. They're really verbose at the moment.

teng-li replied:
Yeah :)

apaszke commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -0,0 +1,196 @@
+       const std::chrono::milliseconds& timeout = kNoTimeout);
+
+// Helper resource guard class
+class ResourceGuard {
```
This is probably useful outside of the `tcputils` context as well. Maybe let's move it out to a separate header?

apaszke commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -0,0 +1,87 @@
+              std::string val = "thread_val_" +
+                  std::to_string(numIterations - 1);
+              c10d::test::check(*slaveStores[i], key, val);
+            }
```
Can you also check the counter in all threads?

teng-li replied:
done

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,326 @@
+}
+
+TCPStoreDaemon::~TCPStoreDaemon() {
+
```
Newline can go

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,326 @@
+}
+
+void TCPStoreDaemon::run() {
+
```
Newline can go

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -0,0 +1,326 @@
+void TCPStoreDaemon::run() {
+
+  // Create the control pipe
+  controlPipeFd_ = std::vector<int>{-1, -1};
```
You know it will be of size 2, so you can use a std::array<int, 2> instead, and initialize in the constructor

teng-li replied:
done

pietern commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -0,0 +1,283 @@
+  return listenPort;
+}
+
+}
```
Add // namespace comment

pietern commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -0,0 +1,196 @@
+  return static_cast<RankType>(rank);
+}
+
+// TCP util namespace
```
Superfluous comment

pietern commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -0,0 +1,87 @@
+  c10d::test::Semaphore sem1, sem2;
+
+  // Each thread will have a slave store to send/recv data
+  std::vector<std::unique_ptr<c10d::TCPStore>> slaveStores;
```
They are not slave stores (replicas or something). Can we stick to daemon/server and worker/client?

teng-li replied:
done

pietern reviewed (commented):
Added a few instances of the style inconsistencies I was talking about in the previous review. There are functions that start with a newline and ones that don't. Also the mixed use of // and /**/ is odd on the eyes.

teng-li committed:
Used notify/wake for wait and addressed all comments

Reference: https://github.com/pytorch/pytorch/issues/7434

pietern commented on torch/lib/c10d/TcpStore.cpp:
```diff
@@ -200,40 +209,55 @@ void TCPStoreDaemon::addHandler(int socket) {
+}
+
+bool TCPStoreDaemon::
+checkKeys(const std::vector<std::string>& keys) const {
```
One line or continuation with argument

pietern reviewed (commented):
Micro nit remaining

teng-li committed:
Clean up nits

apaszke commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -0,0 +1,352 @@
+        // we hit an exception here.
+        ::close(fds[fdIdx].fd);
+        fds.erase(fds.begin() + fdIdx);
+        sockets_.erase(sockets_.begin() + fdIdx - 2);
```
You should remove the socket from the `waiting*` maps as well. Otherwise you'll have the deamon reply to already closed descriptors, and likely crash the whole process.

apaszke replied:
Ok maybe it wouldn't crash the whole process, but it would catch the exception and disconnect the setting process, which is really confusing.

teng-li committed:
Clean up all socket states when the socket is closed

Reference: https://github.com/pytorch/pytorch/issues/7434
